### PR TITLE
Introduce the new notification API in the DM

### PR DESCRIPTION
**Pull request by vincent-zurczak**

### CODE REVIEW

File: core/roboconf-dm-templating/metadata.xml [hunk 1]
```diff
@@ -0,0 +1,27 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<ipojo 
+		xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+		xsi:schemaLocation="org.apache.felix.ipojo http://felix.apache.org/ipojo/schemas/CURRENT/core.xsd"
+		xmlns="org.apache.felix.ipojo">
+
+	<!-- A component for the DM's templating -->
+	<component classname="net.roboconf.dm.templating.internal.TemplatingManager" name="roboconf-dm-templating" public="false">
```
Comment (bourretp) [inline core/roboconf-dm-templating/metadata.xml]: :P Annotations rule!

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/TemplatingManager.java [hunk 2]
```diff
@@ -26,595 +26,347 @@
...
+public class TemplatingManager implements IDmListener {
 
+	public static final String ID = "DM's Templating";
 
-	/**
-	 * The big bad global lock.
-	 */
-	private final ReadWriteLock lock = new ReentrantReadWriteLock(true);
+	// Locks
+	private final Object watcherLock = new Object();
```
Comment (bourretp) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/TemplatingManager.java]: `ReadWriteLock`s allow better throughput, especially when reads vastly outnumber writes (which is the case here).
`synchronized` excludes two concurrent reads :(

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/contexts/ApplicationContextBean.java [hunk 3]
```diff
@@ -54,7 +55,7 @@ public String getDescription() {
 	}
 
 	public Date getLastModified() {
-		return this.lastModified;
+		return new Date( this.lastModified.getTime());
```
Comment (bourretp) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/contexts/ApplicationContextBean.java]: Better! ;)

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+	private static final ThreadFactory THREAD_FACTORY = new WatcherThreadFactory();
+
+	private final Logger logger = Logger.getLogger( getClass().getName());
+	private final AtomicBoolean alreadyStarted = new AtomicBoolean( false );
+
+	private final ReadWriteLock lock = new ReentrantReadWriteLock( true );
+	private final Handlebars handlebars = new Handlebars();
+
+	// FIXME: should we REALLY keep pre-compiled templates in memory?
+	private final Map<File,TemplateEntry> fileToTemplate = new HashMap<> ();
```
Comment (bourretp) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java]: We could use `SoftReference<TemplateEntry>` here, and recompile only when the target object has been (garbage) collected. Very basic caching...

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java [hunk 5]
```diff
@@ -0,0 +1,378 @@
+/**
```
Comment (bourretp) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java]: I think you messed up with git here ;)
File should have been moved, not deleted + new file at new location. Same for TemplateEntry...

File: core/roboconf-dm/src/main/java/net/roboconf/dm/management/Manager.java [hunk 6]
```diff
@@ -379,13 +342,16 @@ public void reconfigure() {
...
 
 		// Restart the templating, using the new configuration directory.
-		synchronized (this) {
-			if( this.templatingManager != null ) {
-				this.templatingManager.stopTemplating();
-				this.templatingManager.startTemplating( this.configurationDirectory );
-				// Applications are re-added as they are reloaded, in the restoreApplications() method.
-			}
-		}
+		// FIXME: will be removed once we implement #394
```
Comment (bourretp) [inline core/roboconf-dm/src/main/java/net/roboconf/dm/management/Manager.java]: ??? Is #394 fixed yet?
AFAIK it isn't, so why are those lines ~~removed~~ commented out?

File: core/roboconf-dm/src/test/java/net/roboconf/dm/management/events/EventTypeTest.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+package net.roboconf.dm.management.events;
+
+import junit.framework.Assert;
+
+import org.junit.Test;
+
+/**
+ * @author Vincent Zurczak - Linagora
+ */
+public class EventTypeTest {
```
Comment (bourretp) [inline core/roboconf-dm/src/test/java/net/roboconf/dm/management/events/EventTypeTest.java]: Why the hell?

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/TemplatingManager.java [hunk 8]
```diff
@@ -26,595 +26,347 @@
...
-	/**
-	 * The templates directory: {@code ${configDir}}/{@value #TEMPLATE_DIRECTORY}.
-	 *
-	 * GuardedBy this.lock
-	 * @see #resetTemplateWatcher()
-	 */
-	private File templateDir;
+	// Fields
+	private final Logger logger = Logger.getLogger( getClass().getName());
+	File templatesDIR, outputDIR;
```
Comment (bourretp) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/TemplatingManager.java]: I have serious doubts about reading/writing those fields without locking.

**Comment by bourretp** [comment]
Some minor issues, and a couple of serious concerns...
Without anyone else to fix & review them, let's go for a merge!

File: core/roboconf-dm-templating/metadata.xml (hunk 1 above)
Comment (vincent-zurczak) [inline core/roboconf-dm-templating/metadata.xml]: What I like about this file is that we have iPojo information in a single place.
No need to introspect Java classes or search for annotations. Finding such information quickly may help to update [this page](http://roboconf.net/en/developer-guide/ipojo-model.html).

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/TemplatingManager.java (hunk 2 above)
Comment (vincent-zurczak) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/TemplatingManager.java]: I created #405 to discuss this in a live meeting.

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java (hunk 5 above)
Comment (vincent-zurczak) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java]: Maybe an issue with EGit (Eclipse's Git implementation).
I just moved the files in my IDE. I did not look at how it appeared in Git commits.

File: core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java (hunk 4 above)
Comment (vincent-zurczak) [inline core/roboconf-dm-templating/src/main/java/net/roboconf/dm/templating/internal/templates/TemplateWatcher.java]: The underlying question is about whether we need to have a `Template` field in **TemplateEntry**. We could create it on the fly when we need it.

File: core/roboconf-dm/src/main/java/net/roboconf/dm/management/Manager.java (hunk 6 above)
Comment (vincent-zurczak) [inline core/roboconf-dm/src/main/java/net/roboconf/dm/management/Manager.java]: Nope, it is not fixed.
But I don't see the interest of working on something I will delete very soon. :game_die:

File: core/roboconf-dm/src/test/java/net/roboconf/dm/management/events/EventTypeTest.java (hunk 7 above)
Comment (vincent-zurczak) [inline core/roboconf-dm/src/test/java/net/roboconf/dm/management/events/EventTypeTest.java]: Indeed...  :pensive:
But for my defense, it took about 30 seconds to write it. Besides, coverage took almost +4% with your merge.

**Pull request merged**